File: db4o.net/Db4objects.Db4o.CS/Db4objects.Db4o.CS/CS/Internal/ClientHeartbeat.cs
/* Copyright (C) 2004 - 2011  Versant Inc.  http://www.db4o.com */

using System;
using Db4objects.Db4o.CS.Internal;
using Db4objects.Db4o.CS.Internal.Messages;
using Db4objects.Db4o.Foundation;
using Db4objects.Db4o.Internal;
using Sharpen.Lang;

namespace Db4objects.Db4o.CS.Internal
{
	/// <exclude></exclude>
	public class ClientHeartbeat : IRunnable
	{
		private SimpleTimer _timer;

		private readonly ClientObjectContainer _container;

		public ClientHeartbeat(ClientObjectContainer container)
		{
			_container = container;
			_timer = new SimpleTimer(this, Frequency(container.ConfigImpl));
		}

		private int Frequency(Config4Impl config)
		{
			return Math.Min(config.TimeoutClientSocket(), config.TimeoutServerSocket()) / 4;
		}

		public virtual void Run()
		{
			_container.WriteMessageToSocket(Msg.Ping);
		}

		public virtual void Start()
		{
			_container.ThreadPool().Start("db4o client heartbeat", _timer);
		}

		public virtual void Stop()
		{
			if (_timer == null)
			{
				return;
			}
			_timer.Stop();
			_timer = null;
		}
	}
}

File: db4o.cs/src/com/db4o/cs/internal/ClientHeartbeat.java
/* Copyright (C) 2007  Versant Inc.  http://www.db4o.com */

package com.db4o.cs.internal;

import com.db4o.cs.internal.messages.*;
import com.db4o.foundation.*;
import com.db4o.internal.*;

/**
 * @exclude
 */
public class ClientHeartbeat implements Runnable {
    
    private SimpleTimer _timer; 
    
    private final ClientObjectContainer _container;

    public ClientHeartbeat(ClientObjectContainer container) {
        _container = container;
        _timer = new SimpleTimer(this, frequency(container.configImpl()));
    }
    
    private int frequency(Config4Impl config){
        return Math.min(config.timeoutClientSocket(), config.timeoutServerSocket()) / 4;
    }

    public void run() {
        _container.writeMessageToSocket(Msg.PING);
    }
    
    public void start(){
    	_container.threadPool().start("db4o client heartbeat", _timer);
    }

    public void stop() {
        if (_timer == null){
            return;
        }
        _timer.stop();
        _timer = null;
    }

}
